### ISSUE TITLE
SpringBoot + Dockers + PinpointAgent 

### ISSUE THREAD

**Issue by manjunathsv1947** [naver/pinpoint#5340]
Sorry for opening duplicate issue, I thought once closed issue won't addressed!   
https://github.com/naver/pinpoint/issues/5338 was closed!

Due to architectural complications of our project setup, We choose not to follow pinpoint-docker steps described by naver for now.
I have 2 machines, both have docker(Docker version 17.09.0-ce, build afdb6d4) running.

**Machine 1 is my Mac.**

container is setup as below
**jar**: /opt/java/microservice/application.jar
**pinpoint**: /opt/java/pinpoint/
I start the application using below command:
java -javaagent:/opt/java/pinpoint/pinpoint-agent-1.8.1/pinpoint-bootstrap-1.8.1.jar -Dpinpoint.agentId=agent1 -Dpinpoint.applicationName=Application1 -jar /opt/java/microservices/application/jar/application.jar

**Result**:Application comes up successfully, But there are some warnings from pinpoint.
Please check 
[machine1_warnings.log](https://github.com/naver/pinpoint/files/2957287/machine1_warnings.log)

I had to set profiler.jdbc.oracle=false to get rid of the above warnings. and Application comes up successfully.

**Machine 2 is a CentOS machine on our datacenter.**

container is setup as below
**jar**: /opt/java/microservice/application.jar
**pinpoint**: /opt/java/pinpoint/
I start the application using below command:
java -javaagent:/opt/java/pinpoint/pinpoint-agent-1.8.1/pinpoint-bootstrap-1.8.1.jar -Dpinpoint.agentId=agent1 -Dpinpoint.applicationName=Application1 -jar /opt/java/microservices/application/jar/application.jar

**Result**:Application fails to start with attached errors
[machine_2_Error_1.log](https://github.com/naver/pinpoint/files/2957289/machine_2_Error_1.log)

I set profiler.jdbc.oracle=false and re-run the application. Application again fails with attached
[machine_2_error_2.log](https://github.com/naver/pinpoint/files/2957292/machine_2_error_2.log)

I am using same docker image in both the machines and the same version of pinpoint-agent.
**note**: If I don't attach the -javaagent(pinpoint) argument. There is no issue!! Application in machine 2 comes up successfully.

Can you please help us understand the issue further?

**Comment by Xylus** [comment]
Hi @manjunathsv1947 
The issue occurring in Machine 1 looks similar to the error in #5261. I'm working on a fix and it should be deployed soon.
This issue should only appear when attaching an agent to an executable jar with dependent libraries nested within them (such as spring boot executable jars and one-jar executable jars) with Cassandra Datastax, Oracle jdbc connector, or okhttp client v2 - which seems to match your case here (Spring Boot + Oracle).
Until then, I think you can use the 1.7.3 agent as a work around.

The error in Machine 2 seems to be different and I am not sure of the exact cause.
I am suspecting a conflict between sleuth and pinpoint instrumentation though if this is the case, the same error should be occurring in Machine 1 as well.
Would it be possible for you to try running without sleuth dependency so that we can at least rule this out?

**Comment by manjunathsv1947** [comment]
Hi @Xylus Thanks for quick reply. These comments help us better. 
That is what I thought, sleuth might be causing the issue. But ignored it for everything working fine on my local machine.
Will go ahead and try without sleuth dependency.

**Comment by manjunathsv1947** [comment]
Hi @Xylus 
ojdbc issue mentioned in
https://github.com/naver/pinpoint/files/2957289/machine_2_Error_1.log is gone now. After we switched to ojdbc6 (https://support.oracle.com/knowledge/Middleware/2196488_1.html).

But errors mentioned in 
https://github.com/naver/pinpoint/files/2957292/machine_2_error_2.log still happen. I took out sleuth dependency and tested, Below is the log file attached.
[error_without_sleuth_dependency.log](https://github.com/naver/pinpoint/files/2965709/error_without_sleuth_dependency.log)

I am attaching controller code that is causing the issue for your reference.
[method.txt](https://github.com/naver/pinpoint/files/2965716/method.txt)

Any help or suggestion would be great to proceed further. Thanks!

**Comment by Xylus** [comment]
@manjunathsv1947 
Thanks for the update. It looks like there's something wrong with the byte code generated from instrumentation, but it'll take a bit more time to analyze all this.
Not being able to reproduce it locally isn't helping either :(
Could you tell us the java version you're using on machine 2 (docker)?
Better yet, if you could post/mail me the whole agent DEBUG log running on machine 2, that would be great. (The *log4j.xml* file is in *$PINPOINT_AGENT/lib* directory.

Also, in the mean time, could you try running the agent with spring bean tracing off? You can turn this off by setting `profiler.spring.beans=false` in *pinpoint.config* file.
Thanks.

**Comment by manjunathsv1947** [comment]
Hi @Xylus 
I took sometime to get back. 
We are using below java version. 
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)

I started application by setting _**profiler.spring.beans=false**_ and the application started successfully. Here is the log file: 
[profiler-spring-beans-false.log](https://github.com/naver/pinpoint/files/2977878/profiler-spring-beans-false.log)

Here is the log file with _**profiler.spring.beans=true**_
[profiler-spring-beans-true.log](https://github.com/naver/pinpoint/files/2977886/profiler-spring-beans-true.log)

What is the effect of setting profiler.spring.beans to **false**?

**Comment by manjunathsv1947** [comment]
Also @Xylus 

I am forced to set profiler.tomcat.enable=false in machine 2, Or the application won't come up. 
If I set profiler.tomcat.enable=true and start application? I see below issue. 
[tomcat-enabled-true.log](https://github.com/naver/pinpoint/files/2979350/tomcat-enabled-true.log)

On my local machine application comes up with profiler.tomcat.enable=false/true (true or false). 
But data is recorded on web-ui only if profiler.tomcat.enable=true
![Screen Shot 2019-03-18 at 10 24 44 PM](https://user-images.githubusercontent.com/36981861/54548617-4b7bb580-49ce-11e9-8d7f-dfe03b0e46be.png)

profiler.tomcat.enable=false, Doesn't record any data.

**Comment by Xylus** [comment]
Thanks for the update @manjunathsv1947 
I've been trying everything to replicate the issue without success :(

Anyway, to answer your questions from the past 2 posts.

> What is the effect of setting profiler.spring.beans to false?

Setting profiler.spring.beans to false tells the agent to not transform Spring beans. This includes your `CSPUserController` class, so the **VerifyError** shouldn't occur.

---

> I am forced to set profiler.tomcat.enable=false in machine 2, Or the application won't come up.

Similarly, setting profiler.tomcat.enable to false in machine 2 would stop the agent from transforming Tomcat related classes (including the `org.apache.catalina.connector.Request` class). Setting this to true will throw a **VerifyError**, which is why your application is not starting up.

---

>On my local machine application comes up with profiler.tomcat.enable=false/true (true or false). But data is recorded on web-ui only if profiler.tomcat.enable=true

Setting profiler.tomcat.enable to false here will stop the agent from transforming Tomcat related classes (see above) so no trace will be generated. This is why no data is shown in your web-ui.
Ultimately, you shouldn't need to set any of the options above to false - but unfortunately, this is not working for machine 2.

---

A couple questions/suggestions:
1. Are you using any special jdk to build your *application.jar* binary?
2. There is an option to dump byte code in the agent's *pinpoint.config* file, which could help us figure out what's going on. 
  If you can re-enable `profiler.spring.beans=true`, set `bytecode.dump.enable`, `bytecode.dump.bytecode`, `bytecode.dump.verify`, `bytecode.dump.asm` to true, and set `bytecode.dump.classlist=com.vmware.user.controller.CSPUserController` in *pinpoint.config*, it will log byte code and transformation dumps into the agent log. Comparing dumps between the agents in machine 1 and machine 2 could be useful. (Please post/send us the dumps so we can help out as well).

**Comment by ManjunathSwamy** [comment]
Hi @Xylus 
I took my time to comeback on this :) 

We took the testing to different environments. The above issue pertains to a single environment only. The same image works fine in all the other environments.

Even a **hello-world** springboot application (containerized) fails @ embedded tomcat initialization in the machine 2 with pinpoint, as mentioned in https://github.com/naver/pinpoint/files/2979350/tomcat-enabled-true.log. But works well with pinpoint on all other environments. 

We will explore on the environment differences and try to understand what exactly is causing the issue. Will keep you updated here. 

Thanks for all your support.
